### PR TITLE
New OTA debugging and logging feature

### PR DESCRIPTION
**Pull request by d00616**
On a Node add '#define MY_DEBUG_OTA 0' and on the Gateway add '#define MY_OTA_LOG_RECEIVER_FEATURE'.

### CODE REVIEW

File: MyConfig.h [hunk 1]
```diff
@@ -48,6 +48,19 @@
 //#define MY_DEBUG
 
 /**
+ * @def MY_DEBUG_OTA
+ * @brief Define MY_DEBUG_OTA to redirect debug prints to given node ID
+ *
+ * This option redirects serial debugging to given node ID, while serial Debugging is disabled.
```
Comment (fallberg) [inline MyConfig.h]: I find the line a bit confusing. "while serial  debugging is disabled". You mean to say that when this flag is set, you get no serial debug, instead all debug goes OTA to specified node id, right?

File: MyConfig.h [hunk 2]
```diff
@@ -80,7 +93,7 @@
  */
 //#define MY_SPLASH_SCREEN_DISABLED
 
-#ifdef MY_DISABLED_SERIAL
+#if defined(MY_DISABLED_SERIAL) and !defined(MY_DEBUG_OTA)
```
Comment (fallberg) [inline MyConfig.h]: Is "and" really allowed and portable? Should it not be &&?

File: core/MyOTADebug.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ */
+#ifdef MY_DEBUG_OTA
+
+#include "MyOTADebug.h"
+
+// global variables
+MyMessage msgOTADebug(OTA_DEBUG_CHILD_NODE_ID, V_TEXT);
+bool inOTADebugPrint = false;
```
Comment (fallberg) [inline core/MyOTADebug.cpp]: Could be static

File: core/MyOTADebug.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+void OTADebugPrint(const char *fmt, ... )
+{
+	// Avoid recursion
+	if (inOTADebugPrint==true) {
+		return;
+	}
+	inOTADebugPrint = true;
+
+	//int bufferFree = MY_SERIAL_OUTPUT_SIZE;
```
Comment (fallberg) [inline core/MyOTADebug.cpp]: Leftover?

File: MyConfig.h (hunk 1 above)
Comment (d00616) [reply MyConfig.h]: Right.

File: core/MyOTADebug.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+
+#ifndef MyOTADebug_h
+#define MyOTAFDebug_h
+
+#include "MySensorsCore.h"
+
+#define OTA_DEBUG_NODE_ID (MY_DEBUG_OTA)
+#define OTA_DEBUG_CHILD_NODE_ID (254)
+#define OTA_DEBUG_SEND_ACK (true)
```
Comment (d00616) [inline core/MyOTADebug.h]: @fallberg Should I add the ACK option to MyConfig.h? I think about MY_DEBUG_OTA_ENABLE_ACK.

File: core/MyOTADebug.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	// Send package
+	msgOTADebug.setDestination(OTA_DEBUG_NODE_ID);
+	for (int pos = 0; pos<n; pos+=MAX_PAYLOAD) {
+		uint8_t length = strlen(&fmtBuffer[pos]);
+		if (length>MAX_PAYLOAD) {
+			length = MAX_PAYLOAD;
+		}
+		Serial.print(">>>");
```
Comment (tekka007) [inline core/MyOTADebug.cpp]: MY_SERIALDEVICE should be used here and for all subsequent Serial.

File: core/MyOTADebug.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	OTADebugBufferNode = 255;
+}
+
+inline void OTADebugReceive(const MyMessage &message)
+{
+	// debug message?
+	if ((message.sensor == OTA_DEBUG_CHILD_NODE_ID) && (message.type==OTA_DEBUG_MSG_TYPE)) {
+
+		// Ignore debug messages via broadcast
+		if (message.sender == 255) {
```
Comment (tekka007) [inline core/MyOTADebug.cpp]: use BROADCAST_ADDRESS instead of 255

**Comment by fallberg** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Don't merge. I add OTA logging capabilities with OTA debugging as an option.

**Comment by d00616** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Jenkins, build examples please

**Comment by d00616** [comment]
Jenkins, build examples please

**Pull request merged**